### ISSUE TITLE
What the hey is happening to my path?!

### ISSUE THREAD

**Issue by celsoaguiar** [fabricjs/fabric.js#6531]
Folks, in the canvas below I have an image and a lasso path. When I scroll the image my path gets hidden. I'm doing an amount of scaleY to the path and I suspect that is throwing off visibility after a certain threshold. If I allow the controls to be shown, they don't disappear but the path does. Any idea how that can be fixed?
Anticipated gratitude and hat tips for your help!

![ezgif com-optimize](https://user-images.githubusercontent.com/6743135/90675113-e6557080-e20e-11ea-8671-5eff5a3675a3.gif)

**Comment by melchiar** [comment]
Can you provide a fiddle demonstrating the issue? Without seeing your code there isn't really much we can do to help.

**Comment by celsoaguiar** [comment]
there's some complexity to the code at this point, but I will try to get something working

**Comment by celsoaguiar** [comment]
the code doing the scaling:
```
      this.labelPaths.map(fabricObj => {
        fabricObj.set({
          top: (fabricObj.top === 0) ? 0 : fabricObj.aCoords.tr.y * this.scaleYRatio,
          scaleY: this.scaleYRatio,
        });
      });
```

**Comment by celsoaguiar** [comment]
Ok, I got a fiddle working, but here's the deal: If I use latest version of Fabric 3.6.3, the problem is present. If I switch to using an older version 1.7.3, the issue is gone. You can try that in the fiddle itself. The code presents an image and a rect on a canvas. In 2 secs, the image and and rect get scaled by .2 ratio. After that you can use the middle mouse button to scroll. Scroll down and you'll see that the rect disappears after some point. that doesn't happen if using version 1.7.3 of fabric. thanks!

https://jsfiddle.net/celsoaguiar/9sur1wqj/10/

**Comment by melchiar** [comment]
Hmm...I'm unable to replicate. What browser are you running? (I've tested with chrome and firefox on pc).

**Comment by celsoaguiar** [comment]
Chrome. On mac you have to use the 2 fingers pad scroll. definitely happening on Chrome w/ the fiddle I did. just tried firefox and get the same issue. image and rect are about close to the bottom of the canvas when it happens

**Comment by celsoaguiar** [comment]
I'm on mac, btw

**Comment by melchiar** [comment]
I don't have access to a Mac so I'm not able to test, but it could have something to do with retinaScaling due to the high dpi display. You can try disabling it to see if the issue persists.

**Comment by celsoaguiar** [comment]
Still happens outside retina settings. Did you try both 3.6.3 and 1.7.3 and both work fine on chrome?

**Comment by celsoaguiar** [comment]
@melchiar, you are probably doing something wrong. I just tried on a win machine chrome and the issue happens w/ that same fiddle

**Comment by melchiar** [comment]
here's a gif of what I'm seeing. let me know if I'm misunderstanding your instructions.

![animation](https://user-images.githubusercontent.com/9057412/90717809-51ca2d00-e265-11ea-9265-fd8412295cbc.gif)

**Comment by celsoaguiar** [comment]
yeah, what you are doing looks right. I'm baffled at what could be causing this. btw, you are using 3.6.3 in the fiddle, right? not 1.7.3

**Comment by melchiar** [comment]
I just tried it again and now I'm seeing what you're seeing. it looks like the issue doesn't occur if the object has been selected/deselected. interesting...

**Comment by celsoaguiar** [comment]
selected/deselected, like how?

**Comment by celsoaguiar** [comment]
the issue actually happens even if you remove the image from the fiddle, it's all to do w/ the 'scaleY' and 'top' props of the rect or path in my original

**Comment by celsoaguiar** [comment]
thanks for trying again, at least I know I'm not going crazy here

**Comment by asturur** [comment]
The issue is because every object gets rendered only if it is on the canvas viewport.
Changing with code the rect left/top/scale/size, will change its position in the canvas, without the canvas really knowing it.
To update its status `legally` you should call `rect.setCoords()`.

Once the object gets rendered as selected, or moved with user interaction, setCoords() get called because is needed to draw controls in place.

setCoords is not called automatically because is ok to call it for one object, while is not ok to call it every render for every object on screen ( calculation have been reduced over time, but with the api control can go up in number if you have many controls ).

So add rect.setCoords() after you change rect position properties, and the problem goes away.

Maybe we should add to this page that another symptom of misaligned coords is that the objects goes offscreen in the wrong moment.

http://fabricjs.com/fabric-gotchas

**Comment by celsoaguiar** [comment]
I get it now, thanks @asturur. Unfortunately, that also means I cannot imitate what Paper.js does in "Scaling an Item Around a Center Point"  (the 3rd demo at http://paperjs.org/tutorials/project-items/transforming-items). In my case I want to scale in reference to the top of the canvas, and in order to do that I'll have to keep tabs of the original 'top' for every object I try to scale. Will probably have to keep 'original-top' as a custom property inside the path object.

**Comment by celsoaguiar** [comment]
got it, will be more careful next time. thanks! 
On the subject at hand, I ended up using:
```
      const sel = new fabric.ActiveSelection(this.canvas.getObjects());
      sel.set({ scaleY: this.scaleYRatio });
      sel.destroy();
```
but I am realizing now that setting a new scaleY on a selection will multiply previous scaleY w the new ratio. in other words, if my selection object's scaleY was 0.8 and I set it to 0.7, my resulting object scaleY will be 08*0.7 which is 0.56. Is that how it's supposed to work?
